pkg/securitycontext/admission/admission.go

```diff
@@ -0,0 +1,284 @@
+		client:  kclient,
+		store:   store,
+	}
+}
+
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	// 1.  find all SCCs the pod has access to:
+	// 2.  Fully resolve each SCC
+	//     1.  determine uid from the SCC's run as user policy
+	//     2.  set the uid on the SCC
+	// 3.  Match the pod's SC to an SCC; for each SCC:
+	//     1.  generate the SC
+	//     2.  validate against the pod's SC
+
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	if !ok {
+		return errors.NewBadRequest("a pod was received, but could not convert the request object.")
+	}
+
+	// get the namespace
+	namespace, err := c.client.Namespaces().Get(a.GetNamespace())
+	if err != nil {
+		return errors.NewBadRequest(err.Error())
+	}
+
+	// all pods are required to have a service account
+	serviceAccount, err := getServiceAccount(c.client, namespace.Name, pod)
+	if err != nil {
+		return errors.NewBadRequest(err.Error())
+	}
+
+	// get all constraints that are usable by this service account
+	matchedConstraints, err := c.getMatchingSecurityContextConstraints(namespace.Name, serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	providers := make(map[string]scc.SecurityContextConstraintsProvider, 0)
+	// set pre-allocated values on constraints
+	for _, constraint := range matchedConstraints {
+		if requiresPreAllocatedUIDRange(constraint) {
+			min, max, err := getPreallocatedUIDRange(namespace)
+			if err != nil {
+				return err
+			}
+			constraint.RunAsUser.UIDRangeMin = min
+			constraint.RunAsUser.UIDRangeMax = max
+		}
+
+		if requiresPreAllocatedSELinuxLevel(constraint) {
+			// if we don't have options create them here so we can add the preallocated settings
+			if constraint.SELinuxContext.SELinuxOptions == nil {
+				constraint.SELinuxContext.SELinuxOptions = &kapi.SELinuxOptions{}
+			}
+
+			level, err := getPreallocatedLevel(namespace)
+			if err != nil {
+				return err
+			}
+			constraint.SELinuxContext.SELinuxOptions.Level = level
+		}
+
+		provider, err := scc.NewSimpleProvider(constraint, c.client)
+		if err != nil {
+			return err
+		}
+		providers[constraint.Name] = provider
+	}
+
+Containers:
+	for i, container := range pod.Spec.Containers {
+		for providerName, provider := range providers {
+			// Create a security context for this provider
+			context, err := provider.CreateSecurityContext(pod, &container)
+			if err != nil {
+				return err
+			}
+
+			// Validate the copy of the container against the security context provider
+			container.SecurityContext = context
+			errs := provider.ValidateSecurityContext(pod, &container)
+			if len(errs) != 0 {
+				continue
+			}
+
+			// Set the security context on actual container
+			pod.Spec.Containers[i].SecurityContext = context
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string, 0)
+			}
+			pod.Annotations[validatedSCCAnnotation] = providerName
```

Comment:
what is making use of this annotation? should you start the `Admit()` method by ensuring it is cleared, so that it can only be set if we get here?